docs/rfcs/20201208_fragmentation.md

```diff
@@ -0,0 +1,303 @@
+
+## Mitigation Design
+
+Here we examine several possible solutions to the fragmentation problem.
+
+We broadly group them into avoidance, allocator modifications and application (redpanda) modifications.
+
+### Avoid
+
+Broadly speaking, the solution here is: *don't try to avoid fragmentation, but rather try to avoid failure even when fragmentation occurs*. The main approach is to simply to _avoid large allocations_.
+
+This is the same approach taken by the Linux kernel (which uses a buddy allocator) for `kmalloc` (the guideline is expect allocators greater than order 4 to fail and for a long time the maximum allocation size was 128 KiB) as well as ScyllaDB (see for example issues [#7364](https://github.com/scylladb/scylla/issues/7364), [#6376](https://github.com/scylladb/scylla/issues/6376), [#108](https://github.com/scylladb/scylla/issues/108), [#1546](https://github.com/scylladb/scylla/issues/1546) and many more).
+
+The main tools to _detect_ such cases are using the large allocation warning emitted by Seastar (not currently configured in Redpanda), and the diagnostics when an allocation fails. 
+
+Once a large allocation is detected it, avoiding it depends on the context (browsing the related ScyllaDB issues shows the large variety of approaches taken). Sometimes it is as simple as using a fragmented data structure (linked list, iobuf, deque) rather than a contiguous one. In other cases it was not expected that such a structure became so large in the first place, so some other adjustment is needed.
+
+#### Drawbacks
+
+A significant downside of the this approach is that is mostly reactive: problems are identified only when they cause a crash or (hopefully) when the warning is detected and acted on. Although the issue may be resolved, significant downtime may already have been incurred.
+
+Another issue is that many allocations occur in third party libraries rather than the core code: although we have the source to all third parties and can patch them, this creates a divergence with upstream, creates a maintenance burden, and so on. Furthermore, it may be harder to avoid large allocations in third party libraries since we won't the same tools available to use as 
+
+#### Recommendations
+
+At least, we should enable the large allocation warning, perhaps at 256 KiB. We can have it enabled in testing with a post-processing phase which determines if the discovered large allocations are from new locations which need to vetted to check how large they can get.
```

Comment:
Is this something that we can enable all the time, even in production?

Reply:
@dotnwat - in its current state I don't think it's suitable since it does a backtrace to the log which is probably useless and comes with a lot of overhead. Furthermore, even time a "large" allocation happens the threshold of what is considered large gets bumped by a factor of 1.7 so we make future stack traces less likely.

For production, I'd envision something like a metric exposed for the number of allocations above some threshold or something like that: cheap to track with no dynamic threshold adjustment as above. In that case I guess it would be useful for diagnosing fragmentation issues: if we get an OOM that seems to be fragmentation related, we can check this metric to try to see if the problem was "fragmentation got worse" or "some allocations got larger".

In the doc, I was thinking more of what we could do in automation to detect this: here the backtrace is useful (the dynamic threshold maybe less so).